### PR TITLE
nnue: distance-based WDL blending

### PR DESCRIPTION
**Pull request by jorgenhanssen**
**WIP**: proof of concept, pending retest after endgame tablebases are implemented.

## Summary

Replaces the fixed WDL blending weight with a per-sample weight that varies based on distance-to-end (dte). Positions near the game end trust the outcome more, positions far from the end trust the eval more.

`w(d) = wdl_start + (wdl_end - wdl_start) * e^(-0.05 * d)`

image

The WDL weight decays exponentially from `wdl_end` (100%) at the game end toward `wdl_start` (10%) for positions far from the end, with a half-life of ~14 plies.

The old `--wdl` flag is replaced by `--wdl_start` (default 0.2) and `--wdl_end` (default 0.8). Setting both to the same value recovers the old fixed behavior.

### Changes

- **Data generation**: Track ply count per game, compute `distance_to_end` per sample, include `dte` column in CSV output
- **Training loss**: Per-sample WDL weight tensor via exponential decay instead of a scalar
- **CLI**: `--wdl_start` and `--wdl_end` replace `--wdl`
- **Self-play**: Enforce 50-move rule in game termination
- **Stats**: Log game length distribution (avg/median/min/max) after generation

### Results

SPRT [0, 10] on 59M samples, 10%/100% dynamic vs 10% fixed:

```
--------------------------------------------------
Results of grail-wdl0.1-1 vs grail-wdl0.1 (10+0.1, 1t, 256MB, UHO_Lichess_4852_v1.epd):
Elo: 6.67 +/- 5.38, nElo: 9.86 +/- 7.94
LOS: 99.25 %, DrawRatio: 43.63 %, PairsRatio: 1.11
Games: 7348, Wins: 2467, Losses: 2326, Draws: 2555, Points: 3744.5 (50.96 %)
Ptnml(0-2): [199, 783, 1603, 856, 233], WL/DD Ratio: 2.50
LLR: 2.95 (100.3%) (-2.94, 2.94) [0.00, 10.00]
--------------------------------------------------
SPRT ([0.00, 10.00]) completed - H1 was accepted
```

### CODE REVIEW

**Comment by jorgenhanssen** [comment]
Ok.. so I generated 184m samples to test, and with a 10-50% slope (grail-a) vs fixed 10% (grail-b) I got:

```
--------------------------------------------------
Results of grail-a vs grail-b (10+0.1, 1t, 256MB, UHO_Lichess_4852_v1.epd):
Elo: -1.97 +/- 8.19, nElo: -2.99 +/- 12.43
LOS: 31.87 %, DrawRatio: 47.07 %, PairsRatio: 0.99
Games: 3000, Wins: 1027, Losses: 1044, Draws: 929, Points: 1491.5 (49.72 %)
Ptnml(0-2): [92, 306, 706, 319, 77], WL/DD Ratio: 3.64
--------------------------------------------------
```

So not the big win I was hoping for... best case neutral. Closing.
